### ISSUE TITLE
mesh generation - explicitly define the interior boundary between two membranes

### ISSUE THREAD

**Issue by purusharths** [spatial-model-editor/spatial-model-editor#885]
For my use-case of SME, I need my program to identify explicitly the boundary between two membranes. An easy way to make such differentiation is to define a line element - with dipafferent material number - between  all the vertices of that boundary.

For example: in this mesh, the material number for the boundary in the interior of the mesh is defined by line element with material number -20.  
<br>
![img](https://i.imgur.com/Sskdxqx.png)
<br/>

Right now, I am trying to add this line manually -Is there any way to do that from within SME itself? It gets a bit complicated to find the convex hull of the points if the mesh itself is not convex ":D

**Comment by purusharths** [comment]
@lkeegan , any update on this? Since it gets really complicated to do so for complex geometries, it would be nice to know if it is possible to do so within sme itself rather than "fixing it in post" ;)

**Comment by lkeegan** [comment]
@purusharths sme does construct these boundary lines as part of the mesh generation but there isn't any functionality to export them, but I guess you can extract this information by iterating over all of the edges in the exported mesh, checking which compartment the two triangles are in that share this edge, and collecting e.g. all the ones that have a triangle from compartment 0 and compartment 1 to get the boundary line between these compartments, as in this notebook:

https://gist.github.com/lkeegan/78e983bd9a7de28fd6f6a76d1248c30a

is that roughly what you are wanting to do?

**Comment by purusharths** [comment]
Ah that's quite an interesting approach! What I used was a sort of line clustering which doens't work for complex shapes - but this is definitely more cleaner.  So it is not possible to embed these extra set of lines in gmsh as physical entity in gmsh? My usual approach is to convert from gmsh to vtk, from vtk to inp, and add material number to the internal boundary in inp. As you can imagine this takes a lot of preprocessing time.

 Unfortunately, it doesn't seem like gmsh and inp have have a same indexing for points (ref: https://heibox.uni-heidelberg.de/d/803107f5993d47afb5d9/) I'm not sure if I am missing something but there doesn't seem to be an obvious offset in indexing difference either.

**Comment by lkeegan** [comment]
I don't have much experience with mesh file formats, but `.inp` is listed as one of the supported types for meshio (https://pypi.org/project/meshio/), so maybe you could modify the notebook to add the boundary lines to the meshio mesh object (I guess by adding a "line" section to `cells` with all of the boundary line elements, and a corresponding boundary line id for each to `cell_data`?), then export to inp?

**Comment by purusharths** [comment]
Yeah this was the first approach I tried, but sadly, I don't think the inp format that I have to use is slimiar to ABAQUS format that is recoginzed by meshio because it doesn't read the points:
```
msh = meshio.read("crypt3.inp")
msh
<meshio mesh object>
  Number of points: 0
  No cells.
```

That's why I was trying to add points mannually but the point index set of gmsh and inp is different so it's not possible to directly write the line element. I suppose a reverse lookup for indices based on points is possible, but it would get messy again.
